siglens/siglens
# Discard stale data

# Description
This fixes an issue that sometimes caused a panic.

In `readDictEnc()`, `SegmentFileReader.deTlv` would increase its size if necessary, but never decrease it. This gets iterated in some spots (e.g., `ApplySearchToMatchFilterDictCsg()`), so it would iterate too far. That might be ok (but wasted CPU) if `deTlv` was copying data; however, it's just referencing data (`sfr.deTlv[w] = buf[soffW:idx]` on line 353 of segreader.go), and `deTlv` can outlive the data it's referencing. So it reads garbage data and occasionally that would cause a panic.

# Testing
The new error checking code in rawchecker.go can help indicate when data is not of the expected form. With that logging active, I was able to reliably get error messages from it with this procedure:
1. Start siglens with no data
2. Use sigclient and ingest data using
```
go run main.go ingest esbulk -t 100_000 -d http://localhost:8081/elastic --processCount 100 -n 1 -b 500 -g benchmark
```
(for some reason, using a high `--processCount` really exacerbated this issue; with `--processCount 1` I didn't see this issue)
3. In the UI, run the query `Monday`. You may have to run the query a few times, or scroll down to fetch more records, but you should see the new error handling log messages in just a few queries.

I also added a unit test

# Checklist:
Before marking your pull request as ready for review, complete the following.
